File: Compiler design/Lab4/Java/src/ru/bmstu/iu9/Compiler.java
package ru.bmstu.iu9;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Compiler {
    private List<Message> messages;
    private Map<String, Integer> nameCodes;
    private List<String> names;

    public Compiler() {
        messages = new ArrayList<Message>();
        nameCodes = new HashMap<String, Integer>();
        names = new ArrayList<String>();
    }

    public int addName(String name) {
        if (nameCodes.containsKey(name))
            return nameCodes.get(name);
        else {
            int code = names.size();
            names.add(name);
            nameCodes.put(name,code);
            return code;
        }
    }

    public boolean errors() {
        return messages.size() > 0;
    }

    public String getName(int code) {
        return names.get(code);
    }

    public void addMessage(boolean isErr, Position c, String text) {
        messages.add(new Message(isErr, text, c));
    }

    public void outputMessages() {
        for (Message m : messages) {
            System.out.print(m.isError ? "Error" : "Warning");
            System.out.print(" " + m.position + ": ");
            System.out.println(m.text);
        }
    }

    public Scanner getScanner(String program) {
        return new Scanner(program, this);
    }
}

File: Compiler design/Lab4/OCaml/Compiler.ml
open Batteries;;

class compiler = object (self)
    
    val mutable messages: Message.message BatList.t = []
    val mutable nameCodes: (string,int) BatHashtbl.t = Hashtbl.create 1000;
    val mutable names: BatUTF8.t BatList.t = []

    method add_name name =
        if BatHashtbl.mem nameCodes name then begin
            BatHashtbl.find nameCodes name;
        end
        else begin
            names <- names@[name];
            BatHashtbl.add nameCodes name ((BatList.length names) - 1);
            (BatList.length names) - 1
        end

    method errors =
        not (BatList.is_empty messages)

    method get_name code =
        BatList.at names code

    method add_message is_err pos text =
        messages <- messages@[new Message.message is_err text pos]

    method output_messages =
        BatList.iter (fun m -> Printf.printf "%s: %s %s" (if m#get_error then "Error" else "Warning") m#get_position#to_string m#get_text) messages

    (* method get_scanner (program: BatUTF8.t) = 
        new Scanner.scanner program self *)

    end;;
  
